File: disjoint_set/c/disjoint_set.c
#include "disjoint_set.h"
#include <stdlib.h>

void disjoint_set_init(DisjointSet* ds, int n)
{
	ds->height = (int*) malloc(sizeof(int) * n);
	ds->parent = (int*) malloc(sizeof(int) * n);
	int i;
	for(i = 0; i < n; i++)
	{
		ds->height[i] = 0;
		ds->parent[i] = i;
	}

	ds->findSet = disjoint_set_find;
	ds->unionSet = disjoint_set_union;
	ds->destroy = disjoint_set_destroy;
}

int disjoint_set_find(DisjointSet* ds, int elem)
{
	if(ds->parent[elem] != elem)
		ds->parent[elem] = disjoint_set_find(ds, ds->parent[elem]);
	return ds->parent[elem];
}

void disjoint_set_link(DisjointSet* ds, int root1, int root2)
{
	if(ds->height[root1] > ds->height[root2])
		ds->parent[root2] = root1;
	else
	{
		if(ds->height[root1] == ds->height[root2])
			ds->height[root2]++;
		ds->parent[root1] = root2;
	}
}

void disjoint_set_union(DisjointSet* ds, int elem1, int elem2)
{
	int root1 = disjoint_set_find(ds, elem1);
	int root2 = disjoint_set_find(ds, elem2);
	if(root1 != root2)
		disjoint_set_link(ds, root1, root2);
}

void disjoint_set_destroy(DisjointSet* ds)
{
	free(ds->height);
	free(ds->parent);
}

File: disjoint_set/go/disjointset.go
package main

// UNTESTED

// DisjointSet is a ADT implemented with union-find algorithm
type DisjointSet struct {
	parent map[int]int
	height map[int]int
}

// NewDisjointSet returns a initialized DisjointSet
func NewDisjointSet() *DisjointSet {
	return &DisjointSet{
		parent: make(map[int]int),
		height: make(map[int]int),
	}
}

// MakeSet adds elem as a set
func (ds *DisjointSet) MakeSet(elem int) {
	ds.parent[elem] = elem
	ds.height[elem] = 1
}

// FindSet finds the set that elem belongs to
func (ds *DisjointSet) FindSet(elem int) int {
	if ds.parent[elem] != elem {
		ds.parent[elem] = ds.FindSet(elem)
	}
	return ds.parent[elem]
}

// UnionSet joins two sets into one
func (ds *DisjointSet) UnionSet(elem1, elem2 int) {
	root1, root2 := ds.FindSet(elem1), ds.FindSet(elem2)
	if root1 != root2 {
		ds.linkSet(root1, root2)
	}
}

func (ds *DisjointSet) linkSet(root1, root2 int) {
	if ds.height[root1] > ds.height[root2] {
		delete(ds.height, root2)
		ds.parent[root2] = root1
	} else {
		if ds.height[root1] == ds.height[root2] {
			ds.height[root2]++
		}
		delete(ds.height, root1)
		ds.parent[root1] = root2
	}
}
